kore/src/Kore/Step/Simplification/SubstitutionSimplifier.hs

```diff
@@ -208,8 +212,23 @@ deduplicateSubstitution
             , Map (UnifiedVariable variable) (TermLike variable)
             )
 deduplicateSubstitution makeAnd' =
-    worker Predicate.makeTruePredicate . Substitution.toMultiMap
+    worker Predicate.makeTruePredicate . checkSetVars . Substitution.toMultiMap
   where
+    checkSetVars checkedMap
+      | Map.null problematicSubmap = checkedMap
+      | otherwise = (error . unlines)
+        [ "found SetVar key:"
+        , show key
+        , "with non-singleton list of assignments as value:"
+        , show value
+        ]
+        where
+            problematicSubmap =
+                Map.filterWithKey
```

Comment:
We should use `Map.foldMapWithKey` here, something like

```.hs
getAny $ Map.foldMapWithKey (\k v -> Data.Monoid.Any $ isSetVar k && isNotSingleton v) ...
```

This way, we don't construct any intermediate `Map`.

Reply:
Oh, ok. I thought we need to save some example of (key, value) pair that fails the check and display it in the error message.

Reply:
Hm, that's a good point, that would be useful in the error message. Can you revert the "remove map filter" change, but update the error message to display all the bad substitutions? Sorry. :sweat_smile: